### DISCUSSION TITLE
RUST客户端报错

### DISCUSSION THREAD

**Discussion by licong87** [zdz/ServerStatus-Rust#224] [General]
您好作者，我用2.3.status.sh脚本部署服务端和客户端都是没有问题，后因为想使用vnstat保留流量数据，所以想在客户端部署RUST版本以支持vnstat，机器是甲骨文ARM,系统是UBUNTU20.04，vnstat版本2.6，配置文件已按照要求更改
目前服务端是用status.sh部署的，客户端改为RUST运行的话
目前报错如下：

```
root@ubuntu:/opt/ServerStatus# bash -ex one-touch.sh ./stat_client -a "grpc://XXXXXXXXXX:9394" -u h1 -p p1 -n
+ set -ex
+ WORKSPACE=/opt/ServerStatus
+ mkdir -p /opt/ServerStatus
+ cd /opt/ServerStatus
+ OS_ARCH=x86_64
++ sed 's/\"//g;s/,//g;s/ //g'
++ awk -F : '{print $2}'
++ head -n 1
++ grep tag_name
++ curl -m 10 -sL https://api.github.com/repos/zdz/ServerStatus-Rust/releases/latest
+ latest_version=v1.7.2
+ wget --no-check-certificate -qO server-x86_64-unknown-linux-musl.zip https://github.com/zdz/ServerStatus-Rust/releases/download/v1.7.2/server-x86_64-unknown-linux-musl.zip
+ wget --no-check-certificate -qO client-x86_64-unknown-linux-musl.zip https://github.com/zdz/ServerStatus-Rust/releases/download/v1.7.2/client-x86_64-unknown-linux-musl.zip
+ unzip -o server-x86_64-unknown-linux-musl.zip
Archive:  server-x86_64-unknown-linux-musl.zip
  inflating: stat_server             
  inflating: stat_server.service     
  inflating: config.toml             
+ unzip -o client-x86_64-unknown-linux-musl.zip
Archive:  client-x86_64-unknown-linux-musl.zip
  inflating: stat_client             
  inflating: stat_client.service     
+ mv -v stat_server.service /etc/systemd/system/stat_server.service
renamed 'stat_server.service' -> '/etc/systemd/system/stat_server.service'
+ mv -v stat_client.service /etc/systemd/system/stat_client.service
renamed 'stat_client.service' -> '/etc/systemd/system/stat_client.service'
+ systemctl daemon-reload
+ systemctl start stat_server
+ systemctl start stat_client
+ systemctl status stat_server
● stat_server.service - ServerStatus-Rust Server
     Loaded: loaded (/etc/systemd/system/stat_server.service; disabled; vendor preset: enabled)
     Active: activating (auto-restart) (Result: exit-code) since Sat 2023-06-17 03:55:12 EDT; 35ms ago
    Process: 6482 ExecStart=/opt/ServerStatus/stat_server -c /opt/ServerStatus/config.toml (code=exited, status=203/EXEC)
   Main PID: 6482 (code=exited, status=203/EXEC)
root@ubuntu:/opt/ServerStatus# 
```

**Comment by zdz** [comment]
手动跑一下, 看看输出是啥
```
/opt/ServerStatus/stat_server -c /opt/ServerStatus/config.toml
或
RUST_BACKTRACE=1 RUST_LOG=trace /opt/ServerStatus/stat_server -c /opt/ServerStatus/config.toml
```

**Reply by licong87** [reply]
root@instance-20230615-2235:~# /opt/ServerStatus/stat_server -c /opt/ServerStatus/config.toml
-bash: /opt/ServerStatus/stat_server: cannot execute binary file: Exec format error

不能执行，不知道是不是命令写的不对？

**Reply by licong87** [reply]
root@instance-20230615-2235:~# RUST_BACKTRACE=1 RUST_LOG=trace /opt/ServerStatus/stat_server -c /opt/ServerStatus/config.toml
-bash: /opt/ServerStatus/stat_server: cannot execute binary file: Exec format error
root@instance-20230615-2235:~# 

RUST的也是一样

**Comment by zdz** [comment]
平台架构下错了，例如 x86 不要用 arm 的包

**Reply by licong87** [reply]
我是ARM架构的，那脚本是否有参数可以让他下ARM的包？

**Comment by zdz** [comment]
![image](https://github.com/zdz/ServerStatus-Rust/assets/152173/16d9fbd2-eb24-47f4-b395-918a1b5caf1e)

**Reply by licong87** [reply]
收到，谢谢~
我先手工替换下，回头再搞脚本，谢谢大佬~

**Reply by licong87** [reply]
root@ubuntu:~# bash -ex one-touch.sh ./stat_client -a "grpc://127.0.0.1:9394" -u h1 -p p1 -n
bash: one-touch.sh: No such file or directory
root@ubuntu:~# cd /opt/ServerStatus
root@ubuntu:/opt/ServerStatus# bash -ex one-touch.sh ./stat_client -a "grpc://127.0.0.1:9394" -u h1 -p p1 -n
+ set -ex
+ WORKSPACE=/opt/ServerStatus
+ mkdir -p /opt/ServerStatus
+ cd /opt/ServerStatus
+ OS_ARCH=aarch64
++ sed 's/\"//g;s/,//g;s/ //g'
++ awk -F : '{print $2}'
++ head -n 1
++ grep tag_name
++ curl -m 10 -sL https://api.github.com/repos/zdz/ServerStatus-Rust/releases/latest
+ latest_version=v1.7.2
+ wget --no-check-certificate -qO server-aarch64-unknown-linux-musl.zip https://github.com/zdz/ServerStatus-Rust/releases/download/v1.7.2/server-aarch64-unknown-linux-musl.zip
+ wget --no-check-certificate -qO client-aarch64-unknown-linux-musl.zip https://github.com/zdz/ServerStatus-Rust/releases/download/v1.7.2/client-aarch64-unknown-linux-musl.zip
+ unzip -o server-aarch64-unknown-linux-musl.zip
Archive:  server-aarch64-unknown-linux-musl.zip
  inflating: stat_server             
  inflating: stat_server.service     
  inflating: config.toml             
+ unzip -o client-aarch64-unknown-linux-musl.zip
Archive:  client-aarch64-unknown-linux-musl.zip
  inflating: stat_client             
  inflating: stat_client.service     
+ mv -v stat_server.service /etc/systemd/system/stat_server.service
renamed 'stat_server.service' -> '/etc/systemd/system/stat_server.service'
+ mv -v stat_client.service /etc/systemd/system/stat_client.service
renamed 'stat_client.service' -> '/etc/systemd/system/stat_client.service'
+ systemctl daemon-reload
+ systemctl start stat_server
+ systemctl start stat_client
+ systemctl status stat_server
● stat_server.service - ServerStatus-Rust Server
     Loaded: loaded (/etc/systemd/system/stat_server.service; enabled; vendor preset: enabled)
     Active: active (running) since Sat 2023-06-17 09:12:40 EDT; 3min 41s ago
   Main PID: 3353 (stat_server)
      Tasks: 5 (limit: 6999)
     Memory: 10.3M
     CGroup: /system.slice/stat_server.service
             └─3353 /opt/ServerStatus/stat_server -c /opt/ServerStatus/config.toml

Jun 17 09:12:40 ubuntu systemd[1]: Started ServerStatus-Rust Server.
Jun 17 09:12:40 ubuntu stat_server[3353]: ✨ stat_server v1.7.2 (f2bbe95, 2023-05-28 04:08:36 UTC, rustc 1.69.0, aarch64-unknown-linux-musl)
Jun 17 09:12:40 ubuntu stat_server[3353]: ✨ run in normal mode, load conf from local file `/opt/ServerStatus/config.toml
Jun 17 09:12:40 ubuntu stat_server[3353]: ✨ admin_user: admin
Jun 17 09:12:40 ubuntu stat_server[3353]: ✨ admin_pass: 2a10c41f-e2bd-4cd9-893e-512873ad6ab4
Jun 17 09:12:40 ubuntu stat_server[3353]: 🚀 listening on grpc://0.0.0.0:9394
Jun 17 09:12:40 ubuntu stat_server[3353]: 🚀 listening on http://0.0.0.0:8080
Jun 17 09:15:47 ubuntu systemd[1]: stat_server.service: Current command vanished from the unit file, execution of the command list won't be resumed.
lines 1-17

更改完以后可以运行了，但是似乎最后一步还是报错~

**Reply by licong87** [reply]
可以成功运行了
root@ubuntu:~# cd /opt/ServerStatus
root@ubuntu:/opt/ServerStatus# bash -ex one-touch.sh ./stat_client -a "grpc://127.0.0.1:9394" -u h1 -p p1 -n
+ set -ex
+ WORKSPACE=/opt/ServerStatus
+ mkdir -p /opt/ServerStatus
+ cd /opt/ServerStatus
+ OS_ARCH=aarch64
++ sed 's/\"//g;s/,//g;s/ //g'
++ awk -F : '{print $2}'
++ head -n 1
++ grep tag_name
++ curl -m 10 -sL https://api.github.com/repos/zdz/ServerStatus-Rust/releases/latest
+ latest_version=v1.7.2
+ wget --no-check-certificate -qO server-aarch64-unknown-linux-musl.zip https://github.com/zdz/ServerStatus-Rust/releases/download/v1.7.2/server-aarch64-unknown-linux-musl.zip
+ wget --no-check-certificate -qO client-aarch64-unknown-linux-musl.zip https://github.com/zdz/ServerStatus-Rust/releases/download/v1.7.2/client-aarch64-unknown-linux-musl.zip
+ unzip -o server-aarch64-unknown-linux-musl.zip
Archive:  server-aarch64-unknown-linux-musl.zip
  inflating: stat_server             
  inflating: stat_server.service     
  inflating: config.toml             
+ unzip -o client-aarch64-unknown-linux-musl.zip
Archive:  client-aarch64-unknown-linux-musl.zip
  inflating: stat_client             
  inflating: stat_client.service     
+ mv -v stat_server.service /etc/systemd/system/stat_server.service
renamed 'stat_server.service' -> '/etc/systemd/system/stat_server.service'
+ mv -v stat_client.service /etc/systemd/system/stat_client.service
renamed 'stat_client.service' -> '/etc/systemd/system/stat_client.service'
+ systemctl daemon-reload
+ systemctl start stat_server
+ systemctl start stat_client
+ systemctl status stat_server
● stat_server.service - ServerStatus-Rust Server
     Loaded: loaded (/etc/systemd/system/stat_server.service; enabled; vendor preset: enabled)
     Active: active (running) since Sat 2023-06-17 09:21:09 EDT; 1min 11s ago
   Main PID: 419 (stat_server)
      Tasks: 5 (limit: 6999)
     Memory: 13.2M
     CGroup: /system.slice/stat_server.service
             └─419 /opt/ServerStatus/stat_server -c /opt/ServerStatus/config.toml

Jun 17 09:21:09 ubuntu systemd[1]: Started ServerStatus-Rust Server.
Jun 17 09:21:10 ubuntu stat_server[419]: ✨ stat_server v1.7.2 (f2bbe95, 2023-05-28 04:08:36 UTC, rustc 1.69.0, aarch64-unknown-linux-musl)
Jun 17 09:21:10 ubuntu stat_server[419]: ✨ run in normal mode, load conf from local file `/opt/ServerStatus/config.toml
Jun 17 09:21:10 ubuntu stat_server[419]: ✨ admin_user: admin
Jun 17 09:21:10 ubuntu stat_server[419]: ✨ admin_pass: 60d9c706-edc5-4f7d-8cad-c253c208a1ad
Jun 17 09:21:10 ubuntu stat_server[419]: 🚀 listening on grpc://0.0.0.0:9394
Jun 17 09:21:10 ubuntu stat_server[419]: 🚀 listening on http://0.0.0.0:8080
+ systemctl status stat_client
● stat_client.service - ServerStatus-Rust Client
     Loaded: loaded (/etc/systemd/system/stat_client.service; disabled; vendor preset: enabled)
     Active: active (running) since Sat 2023-06-17 09:22:20 EDT; 64ms ago
   Main PID: 1183 (stat_client)
      Tasks: 1 (limit: 6999)
     Memory: 596.0K
     CGroup: /system.slice/stat_client.service
             └─1183 /opt/ServerStatus/stat_client -a http://127.0.0.1:8080/report -u h1 -p p1

Jun 17 09:22:20 ubuntu systemd[1]: Started ServerStatus-Rust Client.
root@ubuntu:/opt/ServerStatus# 


但是似乎有一个问题，机器重启以后，服务端获取不到数据，如何添加RUST版客服端开机自启？

**Comment by zdz** [comment]
https://github.com/zdz/ServerStatus-Rust/issues/2

**Reply by licong87** [reply]
总结下，需要 vnstat支持，安装rust客户端的朋友需要在下载完
# wget --no-check-certificate -qO one-touch.sh 'https://raw.githubusercontent.com/zdz/ServerStatus-Rust/master/one-touch.sh'
以后先运行 nano one-touch.sh
把里面x86_64替换为aarch64（# 下载, arm 机器替换 x86_64 为 aarch64 OS_ARCH="x86_64"）
之后再运行（更改完ctrl+X , Y回车）
bash -ex one-touch.sh
就可以部署了~
感谢大佬解惑~

**Reply by Onionbai** [reply]
作者你好，我的小服务器（1900）开启的ipv6，没有ipv4，用脚本安装stat_client，在输入server的ip和账户之后就卡住不动了，手动运行的报错是

root@j1900:/opt/ServerStatus/client# ./stat_client -a "grpc://[ipv6addr]:9394" -u user -p passwd
sys id: xxxxxxxxxxxxxxxxxxxxxxxxx
sys info: {"name":"j1900","version":"1.7.2","os_name":"linux","os_arch":"x86_64","os_family":"unix","os_release":"Linux 20.04 Ubuntu","kernel_version":"5.4.0-150-generic","cpu_num":4,"cpu_brand":"Intel(R) Celeron(R) CPU  J1900  @ 1.99GHz","cpu_vender_id":"GenuineIntel","host_name":"j1900"}
feature native enabled
get_network (ipv4, ipv6) => (false, true)
ERROR stat_client > refresh_ip_info error => error decoding response body: expected value at line 1 column 1
Caused by:
   expected value at line 1 column 1
ERROR stat_client > refresh_ip_info error => error decoding response body: expected value at line 1 column 1
Caused by:
   expected value at line 1 column 1

不晓得为啥，在1900 curl [ipv6addr]:9394，返回的是curl: (1) Received HTTP/0.9 when not allowed，晓得了，是因为grpc服务器。

用python3，执行stat_client.py，也是报错，给出的报错太多了，截了一小段

Traceback (most recent call last):
  File "stat_client.py", line 395, in http_report
    r = sess.post(options.addr, auth=auth,
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 583, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 535, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 648, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 498, in send
    raise ConnectionError(err, request=request)
requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
^CTraceback (most recent call last):
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 665, in urlopen
    httplib_response = self._make_request(
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 421, in _make_request
    six.raise_from(e, None)
  File "<string>", line 3, in raise_from
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 416, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.8/http/client.py", line 1348, in getresponse
    response.begin()
  File "/usr/lib/python3.8/http/client.py", line 316, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.8/http/client.py", line 277, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/lib/python3.8/socket.py", line 669, in readinto
    return self._sock.recv_into(b)
ConnectionResetError: [Errno 104] Connection reset by peer

这个Connection reset by peer，我测试了从1900到服务器，ipv6是通的。

我找了个把小时也找不出来为啥，请大佬给看看吧

另外，[cppla][ServerStatus](https://github.com/cppla/ServerStatus)版本的serverstatus，我的这个1900是可以正常运行的，但是另几个没法运行😱，在你这个里就是1900这一个没法运行...

**Reply by Onionbai** [reply]
脚本安装过程附上

root@j1900:~$./status.sh -i -c

[信息] 下载 x86_64 二进制文件
[信息] 检查 x86_64 二进制文件
检测到 /root/client-x86_64-unknown-linux-musl.zip 已存在，是否重新下载？ (Y/N): N
[信息] 取消下载 x86_64 二进制文件
Archive:  /root/client-x86_64-unknown-linux-musl.zip
  inflating: stat_client             
  inflating: stat_client.service     
文件下载和解压成功！
[注意] 请输入服务端的信息, 格式为 "protocol://username:password@master:port" (如输入错误 可以重新运行写入配置)  示例："http://h1:p1@127.0.0.1:8080"
grpc://user:passwd@[ipv6addr]:9394
[信息] 使用 grpc 连接

......(光标就在这一直闪，没有动静

**Reply by licong87** [reply]
老哥，你单独开一个dis给作者，这样作者不一定看得到。。。。。。

**Comment by licong87** [comment]
```
root@ubuntu:~# cd /opt/ServerStatus
root@ubuntu:/opt/ServerStatus# bash -ex one-touch.sh ./stat_client -a "grpc://127.0.0.1:9394" -u h1 -p p1 -n
+ set -ex
+ WORKSPACE=/opt/ServerStatus
+ mkdir -p /opt/ServerStatus
+ cd /opt/ServerStatus
+ OS_ARCH=aarch64
++ sed 's/\"//g;s/,//g;s/ //g'
++ awk -F : '{print $2}'
++ head -n 1
++ grep tag_name
++ curl -m 10 -sL https://api.github.com/repos/zdz/ServerStatus-Rust/releases/latest
+ latest_version=v1.7.2
+ wget --no-check-certificate -qO server-aarch64-unknown-linux-musl.zip https://github.com/zdz/ServerStatus-Rust/releases/download/v1.7.2/server-aarch64-unknown-linux-musl.zip
+ wget --no-check-certificate -qO client-aarch64-unknown-linux-musl.zip https://github.com/zdz/ServerStatus-Rust/releases/download/v1.7.2/client-aarch64-unknown-linux-musl.zip
+ unzip -o server-aarch64-unknown-linux-musl.zip
Archive:  server-aarch64-unknown-linux-musl.zip
  inflating: stat_server             
  inflating: stat_server.service     
  inflating: config.toml             
+ unzip -o client-aarch64-unknown-linux-musl.zip
Archive:  client-aarch64-unknown-linux-musl.zip
  inflating: stat_client             
  inflating: stat_client.service     
+ mv -v stat_server.service /etc/systemd/system/stat_server.service
renamed 'stat_server.service' -> '/etc/systemd/system/stat_server.service'
+ mv -v stat_client.service /etc/systemd/system/stat_client.service
renamed 'stat_client.service' -> '/etc/systemd/system/stat_client.service'
+ systemctl daemon-reload
+ systemctl start stat_server
+ systemctl start stat_client
+ systemctl status stat_server
● stat_server.service - ServerStatus-Rust Server
     Loaded: loaded (/etc/systemd/system/stat_server.service; disabled; vendor preset: enabled)
     Active: active (running) since Sat 2023-06-17 09:52:18 EDT; 3min 14s ago
   Main PID: 2727 (stat_server)
      Tasks: 5 (limit: 6999)
     Memory: 10.3M
     CGroup: /system.slice/stat_server.service
             └─2727 /opt/ServerStatus/stat_server -c /opt/ServerStatus/config.toml

Jun 17 09:52:18 ubuntu systemd[1]: Started ServerStatus-Rust Server.
Jun 17 09:52:18 ubuntu stat_server[2727]: ✨ stat_server v1.7.2 (f2bbe95, 2023-05-28 04:08:36 UTC, rustc 1.69.0, aarch64-unknown-linux-musl)
Jun 17 09:52:18 ubuntu stat_server[2727]: ✨ run in normal mode, load conf from local file `/opt/ServerStatus/config.toml
Jun 17 09:52:18 ubuntu stat_server[2727]: ✨ admin_user: admin
Jun 17 09:52:18 ubuntu stat_server[2727]: ✨ admin_pass: 2e0a899d-14fe-4530-bd96-c248402a27f3
Jun 17 09:52:18 ubuntu stat_server[2727]: 🚀 listening on grpc://0.0.0.0:9394
Jun 17 09:52:18 ubuntu stat_server[2727]: 🚀 listening on http://0.0.0.0:8080
+ systemctl status stat_client
● stat_client.service - ServerStatus-Rust Client
     Loaded: loaded (/etc/systemd/system/stat_client.service; disabled; vendor preset: enabled)
     Active: active (running) since Sat 2023-06-17 09:52:18 EDT; 3min 14s ago
   Main PID: 2738 (stat_client)
      Tasks: 7 (limit: 6999)
     Memory: 6.4M
     CGroup: /system.slice/stat_client.service
             └─2738 /opt/ServerStatus/stat_client -a http://127.0.0.1:8080/report -u h1 -p p1

Jun 17 09:52:18 ubuntu systemd[1]: Started ServerStatus-Rust Client.
Jun 17 09:52:18 ubuntu stat_client[2738]: sys id: d36a36d310485a9928cad8acc07c79b3
Jun 17 09:52:18 ubuntu stat_client[2738]: sys info: {"name":"h1","version":"1.7.2","os_name":"linux","os_arch":"aarch64","os_family":"unix","os_release":"Linux 20.04 U>
Jun 17 09:52:18 ubuntu stat_client[2738]: feature native enabled
Jun 17 09:52:19 ubuntu stat_client[2738]: get_network (ipv4, ipv6) => (true, false)
lines 1-14/14 (END)
```

请教下大佬我重新开了机器部署了一下，服务端OK的，但是到客户端这里有报错，虽然目前界面是有显示的，请大佬帮忙看下是哪里的问题，谢谢~
